qlib/data/data.py

```diff
@@ -587,7 +587,10 @@ def dataset_processor(instruments_d, column_names, start_time, end_time, freq, i
 
         if len(new_data) > 0:
             data = pd.concat(new_data, names=["instrument"], sort=False)
-            data = DiskDatasetCache.cache_to_origin_data(data, column_names)
+
+            # NOTE: InstProcessors may add new columns and using cache_to_origin_data will remove those added columns.
+            if not len(inst_processors):
+                data = DiskDatasetCache.cache_to_origin_data(data, column_names)
```

Comment:
Perhaps modifying the cache_to_origin_data without checking len(inst_processors) is better?

```
    def cache_to_origin_data(data, fields):
        """cache data to origin data

        :param data: pd.DataFrame, cache data.
        :param fields: feature fields.
        :return: pd.DataFrame.
        """
        not_space_fields = remove_fields_space(fields)
        data_selected = data.loc[:, not_space_fields]
        # set features fields
        data_selected.columns = [str(i) for i in fields]

        _fields = [col for col in data.columns if col not in not_space_fields]
        _data_selected = data.loc[:, _fields]
        data = pd.concat([data_selected, _data_selected], axis=1)
        return data
```